### DISCUSSION TITLE
How can I accept Server Sent Event use http client using curl.

### DISCUSSION THREAD

**Discussion by chenghuiwen** [curl/curl#13395] [Q&A] [answered]
How can I accept Server Sent Event use http client using curl.

**Comment by jay** [comment]
> How can I accept Server Sent Event use http client using curl.

Please give some detail about what you are trying to do

**Reply by chenghuiwen** [reply]
I post my request to LLM Web through curl and  try to accept LLM text summary result from Web.
Web return results multiple times,but the curl wirte callback function only trigger one time and buffers Include all results.
How should I use curl to accept text summary result multiple times？
![Server Sent Event](https://github.com/curl/curl/assets/19405022/8971cd49-f3c6-4058-b238-a97dcdd1254d)

**Comment by dfandrich** [comment]
You're going to have to provide more technical details. If your program
performs an HTTP request, libcurl will call the write callback repeatedly until
the entire response for that request is returned. If the program then performs
another request, then its response will be returned likewise until complete. If
the system is using something like server-side events, then the response has
unlimited length but will come in drips and drabs, with each part in a write
callback.

The diagram you show is how libcurl already works, except that the write
callback is not called on every line, but rather whenever there is data 
received, no matter if it's one line, many lines, or only a part of a single
line. If the program wants it line-by-line, then it will need to include a line
parser in the write callback and buffer the responses so it can feed them
to the program line-by-line.

**Reply by chenghuiwen** [reply]
Yeah,but real-time performance is not guaranteed. Only the server replies with all results then the write callback is called repeatedly.Instead of server replies one result and write callback is called  one time.
![curl](https://github.com/curl/curl/assets/19405022/6278a0d4-674f-4439-b3d2-b809bbc0935c)

**Reply by bagder** [reply]
> Yeah,but real-time performance is not guaranteed

curl cannot guarantee something the network and system themselves cannot. But if data arrives, curl will deliver it to the application without delay, in real-time if you will.

A "server-sent event" is just an HTTP response that does not end immediately. Nothing special for curl.

**Reply by chenghuiwen** [reply]
I find curl_easy_perform() is executed in blocking mode, which means that the main program will stop here and wait for libcurl to complete or fail to execute TCP connection, HTTP GET, and HTTP response before continuing to execute. So write callback can‘t trigger.
As shown below, write callback can't transfer TCP data.

**Reply by bagder** [reply]
First, you can use `curl_multi_perform()` if you prefer a non-blocking API.

Then: `curl_easy_perform()` calls the callback when data arrives so yes, it will deliver all the data in a timely manner to the application.

**Reply by chenghuiwen** [reply]
upload image failed. I can accept final http result. But how to accept Intermediate results TCP data，when I post http request.

**Reply by chenghuiwen** [reply]
![Uploading image.png…]()

**Reply by bagder** [reply]
by specifying the write callback to libcurl

**Reply by chenghuiwen** [reply]
I set wirte callback curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &OnWriteData); and OnWriteData trigger when HTTP response result arrive and can't trigger when TCP response data arrive.
TCP data
50	7.089244	10.69.189.113	10.68.237.162	TCP	75	8082 → 33024 [PSH, ACK] Seq=939 Ack=1664 Win=32384 Len=7 TSval=3000135387 TSecr=2698024905 [TCP segment of a reassembled PDU]
![TCP](https://github.com/curl/curl/assets/19405022/a1a444fd-dd60-4616-a790-516e6883b354)

**Reply by bagder** [reply]
If curl did not deliver any data to the application as a result of the incoming TCP segment, then presumably there was no piece of HTTP payload in there. Or perhaps it was just a piece of an HTTP header, as curl will not deliver partial headers to its callback.

**Reply by chenghuiwen** [reply]
Is there any other way for curl to deliver TCP segment to its callback for above TCP data?

**Reply by bagder** [reply, marked as answer]
curl delivers data for the protocols it supports, not raw TCP. You talk about HTTP server events here so its not really clear how getting raw TCP data would help.

If you want raw TCP, just open a socket and connect to a server.

**Reply by chenghuiwen** [reply]
I got it, I try to  use other ways.

**Comment by chenghuiwen** [comment]
In fact, I want to chat with openai(https://platform.openai.com/docs/api-reference/chat/create) using libcur(C++). But I don't know how to real-time accept data-only server-sent events.
